File: src/mbforge/agent/context.py
"""分层对话上下文管理.

解决简单 message 列表的问题：
- 不同层级的消息有不同的生命周期和优先级
- 支持 token 计数裁剪，确保总 token 不超限
- 支持持久化到本地文件
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from ..models.base import Message
from ..utils.logger import get_logger

logger = get_logger(__name__)


def estimate_tokens(text: str) -> int:
    """估算 token 数量（简化版：中文 ~1.5 token/字，英文 ~0.25 token/字符）."""
    chinese_chars = sum(1 for c in text if '一' <= c <= '鿿')
    other_chars = len(text) - chinese_chars
    return int(chinese_chars * 1.5 + other_chars * 0.25)


@dataclass
class ContextLayer:
    """上下文层."""

    name: str
    messages: list[Message] = field(default_factory=list)
    priority: int = 0  # 优先级：数字越小越重要，越不容易被裁剪
    max_tokens: int = 0  # 0 表示不限制
    ephemeral: bool = False  # True 表示该层消息不会被持久化

    def add(self, role: str, content: str, **kwargs) -> None:
        self.messages.append(Message(role=role, content=content, **kwargs))

    def clear(self) -> None:
        self.messages.clear()

    def token_count(self) -> int:
        """计算本层所有消息的 token 总数."""
        return sum(estimate_tokens(m.content) for m in self.messages)


class LayeredContext:
    """分层上下文管理器.

    四层架构：
    - L0 system:   系统提示（永久保留）
    - L1 project:  项目上下文（当前打开的文件、项目信息等）
    - L2 tools:    工具调用结果（临时，ephemeral）
    - L3 history:  对话历史（可裁剪，按 token 裁剪）
    """

    def __init__(
        self,
        system_prompt: str = "",
        max_history_rounds: int = 20,
        max_total_tokens: int = 32000,
    ):
        self.max_history_rounds = max_history_rounds
        self.max_total_tokens = max_total_tokens

        self._layers: list[ContextLayer] = [
            ContextLayer("system", priority=0),  # L0
            ContextLayer("project", priority=1),  # L1
            ContextLayer("tools", priority=2, ephemeral=True),  # L2
            ContextLayer("history", priority=3),  # L3
        ]

        if system_prompt:
            self.set_system_prompt(system_prompt)

    # ---- 快捷访问 ----

    @property
    def _system(self) -> ContextLayer:
        return self._layers[0]

    @property
    def _project(self) -> ContextLayer:
        return self._layers[1]

    @property
    def _tools(self) -> ContextLayer:
        return self._layers[2]

    @property
    def _history(self) -> ContextLayer:
        return self._layers[3]

    # ---- Token 计算 ----

    def total_token_count(self) -> int:
        """计算所有非 ephemeral 层的 token 总数."""
        return sum(
            layer.token_count()
            for layer in self._layers
            if not layer.ephemeral
        )

    # ---- 系统提示 ----

    def set_system_prompt(self, prompt: str) -> None:
        self._system.messages = [Message(role="system", content=prompt)]

    # ---- 项目上下文 ----

    def set_project_context(self, context: str) -> None:
        """设置项目级上下文（会覆盖旧的）."""
        self._project.messages = [
            Message(role="system", content=f"[项目上下文]\n{context}")
        ]

    def update_project_context(self, context: str) -> None:
        """追加项目上下文."""
        self._project.add("system", f"[项目上下文更新]\n{context}")

    def clear_project_context(self) -> None:
        self._project.clear()

    def inject_memory(self, memory_text: str) -> None:
        """注入用户记忆到项目上下文."""
        if not memory_text:
            return
        self._project.add("system", f"[用户记忆]\n{memory_text}")

    def inject_agent_memory(self, memory_text: str) -> None:
        """注入 Agent 学习记忆到项目上下文."""
        if not memory_text:
            return
        self._project.add("system", f"[Agent 经验]\n{memory_text}")

    def inject_retrieval_trajectory(self, trajectory_text: str) -> None:
        """注入检索轨迹到项目上下文（临时，不持久化）."""
        if not trajectory_text:
            return
        self._tools.add("system", f"[检索轨迹]\n{trajectory_text}")

    # ---- 工具结果 ----

    def add_tool_result(
        self, tool_name: str, result: str, tool_call_id: str = ""
    ) -> None:
        """添加工具调用结果到历史层."""
        self._history.add(
            "tool",
            f"[工具调用结果: {tool_name}]\n{result[:4000]}",
            name=tool_name,
            tool_call_id=tool_call_id,
        )

    def clear_tool_results(self) -> None:
        self._tools.clear()

    # ---- 对话历史 ----

    def add_user_message(self, content: str) -> None:
        self._history.add("user", content)

    def add_assistant_message(
        self, content: str, tool_calls: list[dict] | None = None
    ) -> None:
        self._history.add("assistant", content, tool_calls=tool_calls or None)

    def trim_history(self) -> None:
        """基于 token 计数裁剪对话历史，确保不超过 max_total_tokens."""
        # 先按轮次裁剪
        msgs = self._history.messages
        if len(msgs) > self.max_history_rounds * 2:
            keep = self.max_history_rounds * 2
            self._history.messages = msgs[-keep:]
            logger.debug(f"History trimmed to {keep} messages (round limit)")

        # 再按 token 裁剪
        while self.total_token_count() > self.max_total_tokens and len(self._history.messages) > 2:
            # 移除最早的一轮（user + assistant）
            removed = self._history.messages.pop(0)
            if self._history.messages and self._history.messages[0].role == "assistant":
                self._history.messages.pop(0)
            logger.debug(f"Trimmed message to fit token limit (removed: {removed.content[:50]}...)")

    def clear_history(self) -> None:
        self._history.clear()
        self._tools.clear()

    # ---- 组装消息 ----

    VALID_ROLES = {"system", "user", "assistant", "tool"}

    def build_messages(
        self,
        include_tools: bool = True,
        include_history: bool = True,
    ) -> list[Message]:
        """按优先级组装消息列表，供 LLM 调用."""
        result: list[Message] = []

        # L0: system（必须）
        result.extend(self._system.messages)

        # L1: project
        result.extend(self._project.messages)

        # L2: tools（临时结果）
        if include_tools:
            result.extend(self._tools.messages)

        # L3: history
        if include_history:
            self.trim_history()
            result.extend(self._history.messages)

        # Validate roles
        for msg in result:
            if msg.role not in self.VALID_ROLES:
                raise ValueError(f"Invalid message role: {msg.role}")

        return result

    # ---- 持久化 ----

    def save_to_file(self, path: Path) -> None:
        """保存上下文到本地文件（不含 ephemeral 层）."""
        path.parent.mkdir(parents=True, exist_ok=True)
        data = self.to_dict()
        with open(path, "w", encoding="utf-8") as f:
            json.dump(data, f, ensure_ascii=False, indent=2)
        logger.debug(f"Context saved to {path}")

    @classmethod
    def load_from_file(cls, path: Path) -> LayeredContext | None:
        """从本地文件加载上下文."""
        if not path.exists():
            return None
        try:
            with open(path, encoding="utf-8") as f:
                data = json.load(f)
            return cls.from_dict(data)
        except Exception as e:
            logger.warning(f"Failed to load context from {path}: {e}")
            return None

    def to_dict(self) -> dict[str, Any]:
        """序列化为字典（不含 ephemeral 层）."""
        return {
            "system": [m.__dict__ for m in self._system.messages],
            "project": [m.__dict__ for m in self._project.messages],
            "history": [m.__dict__ for m in self._history.messages],
            "max_history_rounds": self.max_history_rounds,
            "max_total_tokens": self.max_total_tokens,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> LayeredContext:
        ctx = cls(
            max_history_rounds=data.get("max_history_rounds", 20),
            max_total_tokens=data.get("max_total_tokens", 32000),
        )
        ctx._system.messages = [Message(**m) for m in data.get("system", [])]
        ctx._project.messages = [Message(**m) for m in data.get("project", [])]
        ctx._history.messages = [Message(**m) for m in data.get("history", [])]
        return ctx

File: src-tauri/src/core/context.rs
use std::path::Path;

use serde::{Deserialize, Serialize};

use super::helpers::estimate_tokens;

/// A single message in the conversation.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    pub role: String,
    pub content: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_calls: Option<Vec<serde_json::Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_call_id: Option<String>,
}

impl Message {
    pub fn system(content: &str) -> Self {
        Self { role: "system".into(), content: content.to_string(), tool_calls: None, name: None, tool_call_id: None }
    }
    pub fn user(content: &str) -> Self {
        Self { role: "user".into(), content: content.to_string(), tool_calls: None, name: None, tool_call_id: None }
    }
    pub fn assistant(content: &str) -> Self {
        Self { role: "assistant".into(), content: content.to_string(), tool_calls: None, name: None, tool_call_id: None }
    }
    pub fn tool(name: &str, content: &str, call_id: &str) -> Self {
        Self { role: "tool".into(), content: content.to_string(), tool_calls: None, name: Some(name.to_string()), tool_call_id: Some(call_id.to_string()) }
    }
}

/// A layer in the layered context.
///
/// Ephemeral layers are automatically cleared after their content is consumed
/// by [`LayeredContext::build_messages`], ensuring temporary context (e.g.
/// retrieval trajectory, RAG injections) does not leak across conversation
/// turns.
#[derive(Debug, Clone, Default)]
struct ContextLayer {
    messages: Vec<Message>,
    ephemeral: bool,
}

impl ContextLayer {
    fn new(ephemeral: bool) -> Self {
        Self { messages: Vec::new(), ephemeral }
    }

    fn token_count(&self) -> usize {
        self.messages.iter().map(|m| estimate_tokens(&m.content)).sum()
    }

    fn clear(&mut self) {
        self.messages.clear();
    }

    fn is_ephemeral(&self) -> bool {
        self.ephemeral
    }
}

/// Layered conversation context (L0-L3).
///
/// - L0 system:   System prompt (permanent)
/// - L1 project:  Project context
/// - L2 tools:    Tool results (ephemeral)
/// - L3 history:  Conversation history (trimmable)
#[derive(Debug, Clone)]
pub struct LayeredContext {
    system: ContextLayer,
    project: ContextLayer,
    tools: ContextLayer,
    history: ContextLayer,
    max_history_rounds: usize,
    max_total_tokens: usize,
}

impl LayeredContext {
    pub fn new(system_prompt: &str, max_history_rounds: usize, max_total_tokens: usize) -> Self {
        let mut ctx = Self {
            system: ContextLayer::new(false),
            project: ContextLayer::new(false),
            tools: ContextLayer::new(true),
            history: ContextLayer::new(false),
            max_history_rounds,
            max_total_tokens,
        };
        if !system_prompt.is_empty() {
            ctx.set_system_prompt(system_prompt);
        }
        ctx
    }

    pub fn set_system_prompt(&mut self, prompt: &str) {
        self.system.messages = vec![Message::system(prompt)];
    }

    pub fn set_project_context(&mut self, context: &str) {
        self.project.messages = vec![Message::system(&format!("[项目上下文]\n{}", context))];
    }

    pub fn inject_memory(&mut self, memory_text: &str) {
        if !memory_text.is_empty() {
            self.project.messages.push(Message::system(&format!("[用户记忆]\n{}", memory_text)));
        }
    }

    pub fn inject_agent_memory(&mut self, memory_text: &str) {
        if !memory_text.is_empty() {
            self.project.messages.push(Message::system(&format!("[Agent 经验]\n{}", memory_text)));
        }
    }

    pub fn inject_retrieval_trajectory(&mut self, trajectory_text: &str) {
        if !trajectory_text.is_empty() {
            self.tools.messages.push(Message::system(&format!("[检索轨迹]\n{}", trajectory_text)));
        }
    }

    pub fn add_user_message(&mut self, content: &str) {
        self.history.messages.push(Message::user(content));
    }

    pub fn add_assistant_message(&mut self, content: &str) {
        self.history.messages.push(Message::assistant(content));
    }

    pub fn add_assistant_message_with_tool_calls(&mut self, content: &str, tool_calls: &[super::llm::ToolCall]) {
        let tc_values: Vec<serde_json::Value> = tool_calls.iter().map(|tc| {
            serde_json::json!({
                "id": tc.id,
                "function": {
                    "name": tc.name,
                    "arguments": tc.arguments.to_string(),
                }
            })
        }).collect();
        self.history.messages.push(Message {
            role: "assistant".into(),
            content: content.to_string(),
            tool_calls: Some(tc_values),
            name: None,
            tool_call_id: None,
        });
    }

    pub fn add_tool_result(&mut self, tool_name: &str, result: &str, tool_call_id: &str) {
        let truncated = if result.len() > 4000 { &result[..result.floor_char_boundary(4000)] } else { result };
        self.history.messages.push(Message::tool(tool_name, truncated, tool_call_id));
    }

    pub fn clear_tool_results(&mut self) {
        self.tools.clear();
    }

    /// Clear history and all ephemeral layers.
    pub fn clear_history(&mut self) {
        self.history.clear();
        self.tools.clear();
    }

    /// Trim history to fit token limits.
    pub fn trim_history(&mut self) {
        // Round-based trimming
        let max_msgs = self.max_history_rounds * 2;
        if self.history.messages.len() > max_msgs {
            let drain = self.history.messages.len() - max_msgs;
            self.history.messages.drain(..drain);
        }

        // Token-based trimming
        while self.total_token_count() > self.max_total_tokens && self.history.messages.len() > 2 {
            self.history.messages.remove(0);
            if self.history.messages.first().map_or(false, |m| m.role == "assistant") {
                self.history.messages.remove(0);
            }
        }
    }

    pub fn total_token_count(&self) -> usize {
        self.system.token_count() + self.project.token_count() + self.history.token_count()
    }

    /// Get history messages without mutating state (read-only).
    pub fn get_history_messages(&self) -> Vec<Message> {
        self.history.messages.clone()
    }

    /// Build message list for LLM call.
    ///
    /// Ephemeral layers (e.g. `tools`) are included when requested and then
    /// automatically cleared so their content does not persist to the next
    /// turn.
    pub fn build_messages(&mut self, include_tools: bool, include_history: bool) -> Vec<Message> {
        let mut result = Vec::new();
        result.extend(self.system.messages.clone());
        result.extend(self.project.messages.clone());
        if include_tools {
            result.extend(self.tools.messages.clone());
            if self.tools.is_ephemeral() {
                self.tools.clear();
            }
        }
        if include_history {
            self.trim_history();
            result.extend(self.history.messages.clone());
        }
        result
    }

    /// Serialize to JSON-compatible map.
    ///
    /// Note: ephemeral layers are intentionally excluded from serialization
    /// as they are transient per-turn context.
    pub fn to_dict(&self) -> serde_json::Value {
        serde_json::json!({
            "system": self.system.messages,
            "project": self.project.messages,
            "history": self.history.messages,
            "max_history_rounds": self.max_history_rounds,
            "max_total_tokens": self.max_total_tokens,
        })
    }

    /// Deserialize from JSON.
    pub fn from_dict(data: &serde_json::Value) -> Self {
        let mut ctx = Self::new(
            "",
            data["max_history_rounds"].as_u64().unwrap_or(20) as usize,
            data["max_total_tokens"].as_u64().unwrap_or(32000) as usize,
        );
        if let Some(sys) = data["system"].as_array() {
            ctx.system.messages = serde_json::from_value(serde_json::Value::Array(sys.clone())).unwrap_or_default();
        }
        if let Some(proj) = data["project"].as_array() {
            ctx.project.messages = serde_json::from_value(serde_json::Value::Array(proj.clone())).unwrap_or_default();
        }
        if let Some(hist) = data["history"].as_array() {
            ctx.history.messages = serde_json::from_value(serde_json::Value::Array(hist.clone())).unwrap_or_default();
        }
        ctx
    }

    /// Save context to file (excludes ephemeral layers).
    pub fn save_to_file(&self, path: &Path) -> Result<(), Box<dyn std::error::Error>> {
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        let data = self.to_dict();
        let json = serde_json::to_string_pretty(&data)?;
        std::fs::write(path, json)?;
        Ok(())
    }

    /// Load context from file.
    pub fn load_from_file(path: &Path) -> Option<Self> {
        let text = std::fs::read_to_string(path).ok()?;
        let data: serde_json::Value = serde_json::from_str(&text).ok()?;
        Some(Self::from_dict(&data))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_message_roles() {
        let m = Message::system("hello");
        assert_eq!(m.role, "system");
        let m = Message::tool("search", "result", "123");
        assert_eq!(m.role, "tool");
        assert_eq!(m.name, Some("search".into()));
    }

    #[test]
    fn test_layered_context() {
        let mut ctx = LayeredContext::new("You are a helper.", 5, 1000);
        ctx.add_user_message("hi");
        ctx.add_assistant_message("hello");
        let msgs = ctx.build_messages(true, true);
        assert!(msgs.len() >= 3); // system + user + assistant
    }

    #[test]
    fn test_token_estimate() {
        let mut ctx = LayeredContext::new("", 5, 1000);
        ctx.add_user_message("Hello world test message");
        assert!(ctx.total_token_count() > 0);
    }

    #[test]
    fn test_ephemeral_layer_auto_clear() {
        let mut ctx = LayeredContext::new("You are a helper.", 5, 1000);
        ctx.inject_retrieval_trajectory("临时检索结果 A");
        assert_eq!(ctx.tools.messages.len(), 1);

        // First build includes ephemeral content, then auto-clears it.
        let msgs = ctx.build_messages(true, true);
        assert!(msgs.iter().any(|m| m.content.contains("临时检索结果 A")));
        assert_eq!(ctx.tools.messages.len(), 0);

        // Second build no longer contains the cleared ephemeral content.
        let msgs2 = ctx.build_messages(true, false);
        assert!(!msgs2.iter().any(|m| m.content.contains("临时检索结果 A")));
    }

    #[test]
    fn test_ephemeral_skipped_when_include_tools_false() {
        let mut ctx = LayeredContext::new("System.", 5, 1000);
        ctx.inject_retrieval_trajectory("RAG context");
        assert_eq!(ctx.tools.messages.len(), 1);

        // When include_tools is false, ephemeral layer is NOT consumed/cleared.
        let msgs = ctx.build_messages(false, true);
        assert!(!msgs.iter().any(|m| m.content.contains("RAG context")));
        assert_eq!(ctx.tools.messages.len(), 1); // still preserved

        // It is still available for a subsequent build with include_tools=true.
        let msgs2 = ctx.build_messages(true, false);
        assert!(msgs2.iter().any(|m| m.content.contains("RAG context")));
        assert_eq!(ctx.tools.messages.len(), 0); // now cleared
    }
}
